web3.js/src/address-lookup-table-program.ts

```diff
@@ -0,0 +1,270 @@
+import * as BufferLayout from '@solana/buffer-layout';
+
+import * as Layout from './layout';
+import { PublicKey } from './publickey';
+import { SystemProgram } from './system-program';
+import { TransactionInstruction } from './transaction';
+import { encodeData, IInstructionInputData } from './instruction';
+
+export type CreateLookupTableParams = {
+    lookupTable: PublicKey;
+    authority: PublicKey;
+    payer: PublicKey;
+    slot: number;
+    bumpSeed: number;
+};
+
+export type FreezeLookupTableParams = {
+    lookupTable: PublicKey;
+    authority: PublicKey;
+};
+
+export type ExtendLookupTableParams = {
+    lookupTable: PublicKey;
+    authority: PublicKey;
+    payer: PublicKey;
+    addresses: Array<PublicKey>;
+};
+
+export type DeactivateLookupTableParams = {
+    lookupTable: PublicKey;
+    authority: PublicKey;
+};
+
+export type CloseLookupTableParams = {
+    lookupTable: PublicKey;
+    authority: PublicKey;
+    recipient: PublicKey;
+};
+
+export type LookupTableInstructionType =
+    | 'Create'
+    | 'Extend'
+    | 'Close'
+    | 'Freeze'
+    | 'Deactivate';
+
+export type LookupTableInstructionInputData = {
+    Create: IInstructionInputData &
+        Readonly<{
+            slot: number;
+            bumpSeed: number;
+        }>;
+    Freeze: IInstructionInputData;
+    Extend: IInstructionInputData &
+        Readonly<{
+            numberOfAddresses: number;
+            addresses: Array<Uint8Array>;
+        }>;
+    Deactivate: IInstructionInputData;
+    Close: IInstructionInputData;
+};
+
+export const LOOKUP_TABLE_INSTRUCTION_LAYOUTS = Object.freeze({
+    Create: {
+        index: 0,
+        layout: BufferLayout.struct<LookupTableInstructionInputData['Create']>([
+            BufferLayout.u32('instruction'),
+            BufferLayout.nu64('slot'),
+            BufferLayout.u8('bumpSeed'),
+        ]),
+    },
+    Freeze: {
+        index: 1,
+        layout: BufferLayout.struct<LookupTableInstructionInputData['Freeze']>([
+            BufferLayout.u32('instruction'),
+        ]),
+    },
+    Extend: (numberOfAddresses: number) => {
+        return {
+            index: 2,
+            layout: BufferLayout.struct<
+                LookupTableInstructionInputData['Extend']
+            >([
+                BufferLayout.u32('instruction'),
+                BufferLayout.nu64('numberOfAddresses'),
+                BufferLayout.seq(
+                    Layout.publicKey(),
+                    numberOfAddresses,
```

Comment:
I believe you can use `BufferLayout.offset(u64(), -8)` to get buffer layout to use the serialized length in this sequence

Reply:
Ah true, might be worth fixing that upstream at some point. Looks like we use `BufferLayout.offset(BufferLayout.u32(), -8)` in a few other places since the serialized u64's are all little endian. Let's go with that instead

Reply:
@parkerqueen ping just in case you missed this ^

Reply:
@jstarry It fails the tests, throwing `encoding overruns Uint8Array`. I can't seem to fix it atm. Any pointers?

Reply:
I resolved it and explained what was broken here: https://github.com/antematter/solana/pull/1

Reply:
Thanks.